### PR TITLE
Add save page box to templates

### PR DESCRIPTION
**Pull request by danacotoran**
This adds the [save this page component](https://government-frontend.herokuapp.com/component-guide/save_this_page) to various templates in `government-frontend`.
We've introduced a feature flag to enable this feature, which will be enabled in staging and disabled in production.

Also adds a `success_alert` to the top of the application template which will provide a confirmation that the page has been saved (or unsaved, respectively) to the user's saved pages.

-----

https://trello.com/c/mjlS9oaX/811-put-the-button-on-government-frontend-pages
https://trello.com/c/611hsn9y/812-add-the-green-save-a-page-confirmation-box-to-government-frontend

------
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: app/helpers/save_this_page_helper.rb [hunk 1]
```diff
@@ -2,8 +2,8 @@ module SaveThisPageHelper
   attr_reader :page_is_saved, :signed_in, :page_path
 
   PAGE_IS_SAVED_LINK_HREF = {
-    true => "/account/saved-pages/remove?page_path=",
-    false => "/account/saved-pages/add?page_path=",
+    true => "#{Plek.find('frontend')}/account/saved-pages/remove?page_path=",
+    false => "#{Plek.find('frontend')}/account/saved-pages/add?page_path=",
```
Comment (barrucadu) [inline app/helpers/save_this_page_helper.rb:6]: This won't work in production because `Plek.find('frontend')` will give you `frontend.blue.whatever` whereas we'll need it to resolve to `www.whatever`

File: app/helpers/save_this_page_helper.rb [hunk 2]
```diff
@@ -15,6 +15,10 @@ def heading_text(options)
     options[:page_is_saved] ? I18n.t("components.save_this_page.page_was_saved_heading") : I18n.t("components.save_this_page.page_not_saved_heading")
   end
 
+  def saved_pages_link
+    "#{Plek.find('frontend')}/account/saved-pages/"
+  end
```
Comment (barrucadu) [inline app/helpers/save_this_page_helper.rb:20]: Similarly here

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -20,4 +22,26 @@ def content_item_path
...
+
+  def save_this_page_enabled?
+    ENV["FEATURE_FLAG_SAVE_A_PAGE"] == "enabled"
+  end
+  helper_method :save_this_page_enabled?
+
+  def user_signed_in?
+    logged_in?
+  end
+  helper_method :user_signed_in?
```
Comment (barrucadu) [inline app/controllers/application_controller.rb:34]: Why use this instead of just `logged_in?`

Comment (huwd) [reply app/controllers/application_controller.rb:34]: Ah right, yeah I had this doing some other stuff when it was in the helper proper.
But your right I could just make a method from the concern a helper_method.

I'll do that now.

File: Gemfile.lock [hunk 4]
```diff
@@ -65,39 +65,39 @@ GEM
     archive-zip (0.12.0)
       io-like (~> 0.3.0)
     ast (2.4.2)
-    better_errors (2.7.1)
+    better_errors (2.9.1)
```
Comment (huwd) [inline Gemfile.lock:68]: There was a conflict between the Gemfile.lock on Main and this one.
I know this is slightly dramatic, but I figured it'd just be best to rebase, delete ours and then re-bundle install.

Along the way we seem to have had updates. Most are minor with the exception of `regexp_parser`.
And tests continue to pass afterwards.

Any objections to bundling like this?
I can separate out into a prior commit or reduce the number of updates if folks prefer/

Comment (AlanGabbianelli) [reply Gemfile.lock:68]: That's what I've done in previous PRs too. As long as nothing breaks, I don't see any problem with that approach.

Comment (huwd) [reply Gemfile.lock]: getting very odd errors on CI with it saying it can't find rake to precompile assets... 😬 
I'm going to make changes here more minimal to isolate what's going on.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 class ApplicationController < ActionController::Base
+  include GovukPersonalisation::AccountConcern
```
Comment (barrucadu) [inline app/controllers/application_controller.rb:2]: Can we instead put all these changes in the `ContentItemsController`?

File: app/helpers/save_this_page_helper.rb [hunk 6]
```diff
@@ -15,6 +15,10 @@ def heading_text(options)
     options[:page_is_saved] ? I18n.t("components.save_this_page.page_was_saved_heading") : I18n.t("components.save_this_page.page_not_saved_heading")
   end
 
+  def saved_pages_link
+    "#{Plek.new.website_root}/account/saved-pages/"
```
Comment (barrucadu) [inline app/helpers/save_this_page_helper.rb:19]: Nitpick: drop the trailing slash.  We don't usually display them for GOV.UK URLs.

Comment (barrucadu) [reply app/helpers/save_this_page_helper.rb:19]: You could also inline this into `additional_text`

Comment (danacotoran) [reply app/helpers/save_this_page_helper.rb]: Thanks, good shout, have done both

File: app/helpers/save_this_page_helper.rb [hunk 7]
```diff
@@ -15,6 +15,10 @@ def heading_text(options)
     options[:page_is_saved] ? I18n.t("components.save_this_page.page_was_saved_heading") : I18n.t("components.save_this_page.page_not_saved_heading")
   end
 
+  def saved_pages_link
+    "#{Plek.new.website_root}/account/saved-pages"
+  end
```
Comment (barrucadu) [inline app/helpers/save_this_page_helper.rb:20]: This method isn't used anywhere any more, right?

File: app/controllers/content_items_controller.rb [hunk 8]
```diff
@@ -199,4 +200,23 @@ def error_redirect(exception)
...
+  end
+  helper_method :save_this_page_enabled?
+
+  helper_method :logged_in?
+
+  def user_has_saved_page?
+    GdsApi.account_api.get_saved_page(
+      page_path: request.path,
+      govuk_account_session: @account_session_header,
+    ).present?
```
Comment (barrucadu) [inline app/controllers/content_items_controller.rb:215]: You'll want to check the `"saved_page"` field: https://github.com/alphagov/account-api/blob/main/app/controllers/saved_pages_controller.rb#L13

File: app/helpers/save_this_page_helper.rb (hunk 7 above)
Comment (danacotoran) [reply app/helpers/save_this_page_helper.rb]: eugh you're right, thought I'd removed it 🤦🏻‍♀️

File: app/helpers/save_this_page_helper.rb [hunk 9]
```diff
@@ -27,6 +27,7 @@ def link_href(options)
 
   def additional_text(options)
     signed_in = options[:signed_in] || false
-    I18n.t(SEE_SAVED_PAGES_LOGGED_IN[signed_in], link: "/account/saved-pages")
+    additional_class = options[:additional_class] || ""
+    I18n.t(SEE_SAVED_PAGES_LOGGED_IN[signed_in], link: "#{Plek.new.website_root}/account/saved-pages", additional_class: additional_class)
```
Comment (barrucadu) [inline app/helpers/save_this_page_helper.rb:31]: Is this meaningfully different to

    I18n.t(SEE_SAVED_PAGES_LOGGED_IN[signed_in], link: "#{Plek.new.website_root}/account/saved-pages", additional_class: options[:additional_class])

?

File: app/views/content_items/working_group.html.erb [hunk 10]
```diff
@@ -35,5 +35,11 @@
         <%= raw(@additional_body) %>
       <% end %>
     <% end %>
+
+    <%= render 'components/save_this_page', {
```
Comment (barrucadu) [inline app/views/content_items/working_group.html.erb:39]: Should this be using the partial instead?

File: app/views/content_items/corporate_information_page.html.erb [hunk 11]
```diff
@@ -56,3 +56,14 @@
     </div>
   </div>
 </div>
+<% if save_this_page_enabled? %>
```
Comment (barrucadu) [inline app/views/content_items/corporate_information_page.html.erb:59]: This too

File: app/helpers/save_this_page_helper.rb (hunk 9 above)
Comment (danacotoran) [reply app/helpers/save_this_page_helper.rb]: I thought the line above was necessary but seems like it's not!

File: app/views/content_items/corporate_information_page.html.erb (hunk 11 above)
Comment (danacotoran) [reply app/views/content_items/corporate_information_page.html.erb:59]: Done! Just as a note, the reason I left the `if save_this_page_enabled?` conditional in here is because we don't want empty divs with govuk grid layout classes in the markup when the save page box is not present.

File: app/controllers/content_items_controller.rb (hunk 8 above)
Comment (huwd) [reply app/controllers/content_items_controller.rb:215]: Was a bit lazy here, but only becuase it 404s / 401s if something goes wrong.

So a sucessful query here by definition has proved presense?
So we can return true.

But sure will adjust this!

**Review by barrucadu — COMMENTED** [review]
Looks pretty good now, thanks for the changes.  I think the last thing I'd want to see is a request spec (or controller spec, whatever we have in government-frontend) that it all works, eg:

```ruby
RSpec.describe "consultations" do
  before { stub_content_item }

  it "does not show the save-page button" do
    # ...
  end

  context "the save-page feature flag is enabled" do
    around { climate_control }

    it "shows the button in the 'not saved' state" do
      # ...
    end
 
    context "the user is logged in" do
      before { log_user_in }

      it "shows the button in the 'not saved' state" do
        # ...
      end

      context "the user has saved the page" do
        before { stub_saved_page }

        it "shows the button in the 'saved' state" do
          # ...
        end
      end
    end
  end
end
```

We've got tests for the component and for the helper but nothing at the level of making a request to the app.

Happy to approve when we've got this in place.

File: app/views/content_items/case_study.html.erb [hunk 12]
```diff
@@ -18,26 +18,30 @@
...
-          alt: @content_item.image["alt_text"],
-          credit: @content_item.image["credit"],
-          caption: @content_item.image["caption"] if @content_item.image %>
-
-      <%= render 'govuk_publishing_components/components/govspeak', {
-        direction: page_text_direction,
-      } do %>
+  <div class="govuk-grid-column-two-thirds">
+    <div class="content-bottom-margin">
+      <div class="responsive-bottom-margin">
```
Comment (andysellick) [inline app/views/content_items/case_study.html.erb:23]: There's a lot of margin being applied around these elements. Something to consider for future tidying perhaps is to have the margin on the components themselves. This might require more work, so non-blocking - the existing margin framework might need to be added to the components being used here, but it would make the markup cleaner and the code more readable.

File: app/views/content_items/case_study.html.erb [hunk 13]
```diff
@@ -18,26 +18,30 @@
...
+        <%= render 'components/figure',
+        src: @content_item.image["url"],
```
Comment (andysellick) [inline app/views/content_items/case_study.html.erb:25]: Nitpick: indentation.

File: app/views/content_items/case_study.html.erb [hunk 14]
```diff
@@ -18,26 +18,30 @@
...
+        credit: @content_item.image["credit"],
+        caption: @content_item.image["caption"] if @content_item.image %>
+
+        <%= render 'govuk_publishing_components/components/govspeak', {
+          direction: page_text_direction,
+        } do %>
         <%= raw(@content_item.body) %>
-      <% end %>
-    </div>
+        <% end %>
```
Comment (andysellick) [inline app/views/content_items/case_study.html.erb:34]: Nitpick: indentation on line 33.

File: app/views/content_items/fatality_notice.html.erb [hunk 15]
```diff
@@ -18,27 +18,29 @@
...
-      } do %>
+        <%= render 'govuk_publishing_components/components/govspeak', {
+          direction: page_text_direction,
+        } do %>
         <%= raw(@content_item.body) %>
-      <% end %>
-    </div>
-
-    <%= render 'components/published-dates', {
+        <% end %>
```
Comment (andysellick) [inline app/views/content_items/fatality_notice.html.erb:35]: Nitpick: indentation on line 34.

File: app/views/content_items/news_article.html.erb [hunk 16]
```diff
@@ -19,33 +19,37 @@
...
-          credit: @content_item.image["credit"],
-          caption: @content_item.image["caption"] if @content_item.image %>
-      <%= render "govuk_publishing_components/components/govspeak", {
-        direction: page_text_direction,
-      } do %>
+  <div class="govuk-grid-column-two-thirds ">
+    <div class="content-bottom-margin">
+      <div class="responsive-bottom-margin">
+        <%= render 'components/figure',
+        src: @content_item.image["url"],
```
Comment (andysellick) [inline app/views/content_items/news_article.html.erb:26]: Nitpick: indentation.

File: app/views/content_items/news_article.html.erb [hunk 17]
```diff
@@ -19,33 +19,37 @@
...
+        alt: @content_item.image["alt_text"],
+        credit: @content_item.image["credit"],
+        caption: @content_item.image["caption"] if @content_item.image %>
+        <%= render "govuk_publishing_components/components/govspeak", {
+          direction: page_text_direction,
+        } do %>
         <%= raw(@content_item.body) %>
-      <% end %>
-    </div>
+        <% end %>
```
Comment (andysellick) [inline app/views/content_items/news_article.html.erb:34]: Nitpick: indentation on line 33.

File: app/views/content_items/news_article.html.erb [hunk 18]
```diff
@@ -19,33 +19,37 @@
...
-    <div class="dont-print responsive-bottom-margin">
-      <%= render 'govuk_publishing_components/components/share_links',
+      <div class="dont-print responsive-bottom-margin">
+        <%= render 'govuk_publishing_components/components/share_links',
         links: @content_item.share_links,
         track_as_sharing: true,
         title: t('components.share_links.share_this_page')
-      %>
-    </div>
+        %>
```
Comment (andysellick) [inline app/views/content_items/news_article.html.erb:42]: Nitpick: indentation lines 39-41.

File: app/views/content_items/publication.html.erb [hunk 19]
```diff
@@ -25,34 +25,37 @@
...
-          mobile_top_margin: true,
-        } %>
-
-        <%= render "govuk_publishing_components/components/govspeak", {
-          direction: page_text_direction,
-        } do %>
+      <%= render 'components/important-metadata', items: @content_item.important_metadata %>
+      <div class="responsive-bottom-margin">
+        <%= render "attachments",
+        title: t("publication.documents", count: 5), # This should always be pluralised.
```
Comment (andysellick) [inline app/views/content_items/publication.html.erb:33]: Nitpick: indentation.

File: app/views/content_items/publication.html.erb [hunk 20]
```diff
@@ -25,34 +25,37 @@
...
+        legacy_pre_rendered_documents: @content_item.documents,
+        attachments: @content_item.featured_attachments %>
+
+
```
Comment (andysellick) [inline app/views/content_items/publication.html.erb:37]: Nitpick: double empty line.

File: app/views/content_items/publication.html.erb [hunk 21]
```diff
@@ -25,34 +25,37 @@
...
+          } %>
+
+          <%= render "govuk_publishing_components/components/govspeak", {
+            direction: page_text_direction,
+          } do %>
           <%= raw(@content_item.details) %>
-        <% end %>
-      </section>
-    </div>
+          <% end %>
```
Comment (andysellick) [inline app/views/content_items/publication.html.erb:48]: Nitpick: indentation on line 47.

File: app/views/content_items/speech.html.erb [hunk 22]
```diff
@@ -19,29 +19,33 @@
...
-
-      <%= render "govuk_publishing_components/components/govspeak", {
+  <div class="govuk-grid-column-two-thirds">
+    <div class="content-bottom-margin">
+      <div class="responsive-bottom-margin">
+        <%= render 'components/important-metadata',
+        items: @content_item.important_metadata %>
+
+        <%= render 'components/figure',
+        src: @content_item.image["url"],
```
Comment (andysellick) [inline app/views/content_items/speech.html.erb:29]: Nitpick: indentation.

File: app/views/content_items/speech.html.erb [hunk 23]
```diff
@@ -19,29 +19,33 @@
...
+        alt: @content_item.image["alt_text"],
+        credit: @content_item.image["credit"],
+        caption: @content_item.image["caption"] if @content_item.image %>
+
+        <%= render "govuk_publishing_components/components/govspeak", {
           direction: page_text_direction,
         } do %>
-          <%= raw(@content_item.body) %>
+        <%= raw(@content_item.body) %>
```
Comment (andysellick) [inline app/views/content_items/speech.html.erb:37]: Nitpick: indentation.

File: app/views/content_items/world_location_news_article.html.erb [hunk 24]
```diff
@@ -19,34 +19,38 @@
...
-          caption: @content_item.image["caption"] if @content_item.image %>
-
-      <%= render 'govuk_publishing_components/components/govspeak', {
-        direction: page_text_direction,
-      } do %>
+  <div class="govuk-grid-column-two-thirds">
+    <div class="content-bottom-margin">
+      <div class="responsive-bottom-margin">
+        <%= render 'components/figure',
+        src: @content_item.image["url"],
```
Comment (andysellick) [inline app/views/content_items/world_location_news_article.html.erb:26]: Nitpick: indentation.

File: app/views/content_items/world_location_news_article.html.erb [hunk 25]
```diff
@@ -19,34 +19,38 @@
...
-      <%= render 'components/figure',
-          src: @content_item.image["url"],
-          alt: @content_item.image["alt_text"],
-          credit: @content_item.image["credit"],
-          caption: @content_item.image["caption"] if @content_item.image %>
-
-      <%= render 'govuk_publishing_components/components/govspeak', {
-        direction: page_text_direction,
-      } do %>
+  <div class="govuk-grid-column-two-thirds">
```
Comment (andysellick) [inline app/views/content_items/world_location_news_article.html.erb:22]: This whole section/general pattern seems to repeat (broadly) on a lot of pages. Would it be possible to consolidate it into a partial?

File: app/views/content_items/world_location_news_article.html.erb [hunk 26]
```diff
@@ -19,34 +19,38 @@
...
+        credit: @content_item.image["credit"],
+        caption: @content_item.image["caption"] if @content_item.image %>
+
+        <%= render 'govuk_publishing_components/components/govspeak', {
+          direction: page_text_direction,
+        } do %>
         <%= raw(@content_item.body) %>
-      <% end %>
-    </div>
+        <% end %>
```
Comment (andysellick) [inline app/views/content_items/world_location_news_article.html.erb:35]: Nitpick: indentation line 34.

File: app/views/layouts/application.html.erb [hunk 27]
```diff
@@ -39,6 +39,12 @@
       <%= render 'govuk_publishing_components/components/phase_banner', phase: @content_item.phase %>
     <% end %>
 
+    <div class="govuk-grid-row">
+      <div class="govuk-grid-column-two-thirds">
+        <%= render 'shared/save_page_confirmation' %>
```
Comment (andysellick) [inline app/views/layouts/application.html.erb:44]: This partial contains logic for whether the contents are shown or hidden, but that doesn't wrap these containing grid divs, which means they'll always be rendered on the page regardless. Would it be better to move them inside the partial, inside the `if` statement?

File: app/views/shared/_save_page_confirmation.html.erb [hunk 28]
```diff
@@ -0,0 +1,11 @@
+<% if has_saved_page_confirmation? %>
+  <div class="govuk-!-margin-top-4">
+    <%= render "govuk_publishing_components/components/success_alert", {
+      message: t("components.save_this_page.confirmations.#{personalisation_param}.message"),
+      description: sanitize(additional_text({
+        :additional_class => "govuk-notification-banner__link",
```
Comment (andysellick) [inline app/views/shared/_save_page_confirmation.html.erb:6]: What's the reason for using this syntax compared with the `additional_class: "govuk-notification-banner__link"` syntax?

File: test/controllers/save_a_page_test.rb [hunk 29]
```diff
@@ -0,0 +1,97 @@
+require "test_helper"
```
Comment (andysellick) [inline test/controllers/save_a_page_test.rb:1]: These tests look good. Does this change also require further tests that the save page functionality works end to end, or will that be in a separate PR?

**Review by andysellick — COMMENTED** [review]
Have given the FE bits a quick look. Mostly indentation points and a few questions, otherwise looks good.

File: app/views/shared/_save_page_confirmation.html.erb (hunk 28 above)
Comment (danacotoran) [reply app/views/shared/_save_page_confirmation.html.erb]: Not sure why I did it like that to be honest 🙃

File: app/views/layouts/application.html.erb (hunk 27 above)
Comment (danacotoran) [reply app/views/layouts/application.html.erb]: Very good point, have done as suggested

File: test/controllers/save_a_page_test.rb [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+  end
+
+  def stub_saved_pages_has_page(page_path)
+    body = { body: { "saved_page": ["some api content here"] }.to_json }
+    stub_saved_pages_query(page_path, 200, body)
+  end
+
+  def stub_saved_pages_no_page(page_path)
+    stub_saved_pages_query(page_path, 404)
+  end
```
Comment (barrucadu) [inline test/controllers/save_a_page_test.rb:74]: Use the test helpers in gds-api-adapters, as if the response format changed the pact tests will start to fail - there's no direct connection between the gds-api-adapters pact tests and the test helpers (unfortunately), but seeing the failing pact tests should prompt someone to look at other places we're making assumptions about the response format.

File: test/controllers/save_a_page_test.rb (hunk 29 above)
Comment (huwd) [reply test/controllers/save_a_page_test.rb:1]: @andysellick and I discssed this.
Outcome: as these are just get links through the frontend controller tests will likely do here.
Frontend is doing all of the fancy API work so we've got fancier pact tests over there!

**Review by huwd — APPROVED** [review]

File: app/helpers/application_helper.rb [hunk 31]
```diff
@@ -28,6 +28,16 @@ def current_path_without_query_string
     request.original_fullpath.split("?", 2).first
   end
 
+  def has_saved_page_confirmation?
+    personalisation_param.present?
+  end
```
Comment (barrucadu) [inline app/helpers/application_helper.rb:33]: One final comment: can you also put this behind the feature flag

Comment (huwd) [reply app/helpers/application_helper.rb:33]: sure!

**Review by barrucadu — APPROVED** [review]
